types/src/bounded_future_queue.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use futures::{
+    stream::{FuturesOrdered, FuturesUnordered},
+    Future, TryFutureExt, TryStreamExt,
+};
+use tokio::sync::{AcquireError, Semaphore, SemaphorePermit};
+
+pub struct UnorderedPermit<'a, T: Future> {
+    permit: SemaphorePermit<'a>,
+    futures: &'a BoundedFuturesUnordered<T>,
+}
+
+/// An async-friendly FuturesUnordered of bounded size. In contrast to a bounded channel,
+/// the queue makes it possible to modify and remove entries in it. In contrast to a FuturesUnordered,
+/// the queue makes it possible to enforce a bound on the number of items in the queue.
+pub struct BoundedFuturesUnordered<T: Future> {
```

Comment:
I believe that probably a good next step (for another PR) on those structures would be to introduce metrics as well to monitor the available capacity as we do with channels - we might want to get notified when becoming saturated . Let me know your thoughts

Reply:
Yep, happy to introduce a metric even in this PR.